File: 2019/5/solution.ts
import * as fs from "fs";
import * as path from "path";
import intcodeGenerator from "../intcode/intcode";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split(",")
  .map(Number);

const intcode = intcodeGenerator(puzzleInput);
const intcode2 = intcodeGenerator(puzzleInput);
intcode.next();
intcode.next(1);
let part1;
while (true) {
  const ret = intcode.next();
  if (ret.done) break;
  part1 = ret.value;
}
console.log(part1);
intcode2.next();
console.log(intcode2.next(5).value);

File: 2019/5/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/2019/intcode"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), ","))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data []string) []int {
	numbers := make([]int, 0, len(data))
	for _, s := range data {
		if n, err := strconv.Atoi(s); err != nil {
			panic(err)
		} else {
			numbers = append(numbers, n)
		}
	}
	return numbers
}

func part1(puzzleInput []int) int {
	ch := make(chan int)
	ic := intcode.NewIntcodeProgram(puzzleInput, ch)
	go ic.Run()
	<-ch
	ch <- 1
	var output int
	for {
		tmp := <-ch
		if tmp == intcode.ENDSIGNAL {
			return output
		}
		output = tmp
	}
}

func part2(puzzleInput []int) int {
	ch := make(chan int)
	ic := intcode.NewIntcodeProgram(puzzleInput, ch)
	go ic.Run()
	<-ch
	ch <- 5
	return <-ch
}
